src/Workspaces/Remote/ServiceHub/Services/LanguageServer.cs

```diff
@@ -0,0 +1,155 @@
+            // our LSP server only supports WorkspaceStreamingSymbolProvider capability
+            // for now
+            return new InitializeResult()
+            {
+                Capabilities = new VSServerCapabilities()
+                {
+                    WorkspaceStreamingSymbolProvider = true
+                }
+            };
+        }
+
+        [JsonRpcMethod(Methods.InitializedName)]
+        public Task Initialized()
+        {
+            return Task.CompletedTask;
+        }
+
+        [JsonRpcMethod(Methods.ShutdownName)]
+        public void Shutdown(CancellationToken cancellationToken)
+        {
+            // our language server shutdown when VS shutdown
+        }
+
+        [JsonRpcMethod(VSSymbolMethods.WorkspaceBeginSymbolName)]
+        public Task<VSBeginSymbolParams> WorkspaceBeginSymbolAsync(string query, int searchId, CancellationToken cancellationToken)
+        {
+            return RunServiceAsync(async () =>
+            {
+                using (UserOperationBooster.Boost())
+                {
+                    // for now, we use whatever solution we have currently. in future, we will add an ability to sync VS's current solution
+                    // on demand from OOP side
+                    await SearchAsync(SolutionService.PrimaryWorkspace.CurrentSolution, query, searchId, cancellationToken).ConfigureAwait(false);
+                    return new VSBeginSymbolParams();
+                }
+            }, cancellationToken);
+        }
+
+        private async Task SearchAsync(Solution solution, string query, int searchId, CancellationToken cancellationToken)
+        {
+            foreach (var project in solution.Projects.Where(p => p.Language == _languageName))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var results = await AbstractNavigateToSearchService.SearchProjectInCurrentProcessAsync(
+                    project,
+                    ImmutableArray<Document>.Empty,
+                    query,
+                    s_supportedKinds,
+                    cancellationToken).ConfigureAwait(false);
+
+                var lspResults = await Convert(results, cancellationToken).ConfigureAwait(false);
+
+                await InvokeAsync(
+                    VSSymbolMethods.WorkspacePublishSymbolName,
+                    new object[] { new VSPublishSymbolParams() { SearchId = searchId, Symbols = lspResults } },
+                    cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private static async Task<VSSymbolInformation[]> Convert(
+            ImmutableArray<INavigateToSearchResult> results, CancellationToken cancellationToken)
+        {
+            var symbols = new VSSymbolInformation[results.Length];
+
+            for (var i = 0; i < results.Length; i++)
+            {
+                var text = await results[i].NavigableItem.Document.GetTextAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
feels expensive.  can we group by document so we only have to do this once per doc?

Reply:
let's see whether this ever show up as perf issue before micro optimizing it.